### DISCUSSION TITLE
How does `GravatarImageSource` access `ImageSource.CancellationTokenSource`?

### DISCUSSION THREAD

**Discussion by Ghostbird** [CommunityToolkit/Maui#1097] [Q&A] [answered]
I was trying to write a `StreamImageSource` to load streamed images from a specific API, and used the `GravatarImageSource` as an example. At some point it delays loading for a few milliseconds and calls `CancellationTokenSource.Cancel()`. It took me a while to get what was happening there. I still don't really get why it's done that way. My questions are:

- Should I do the same in my implementation?
- How can the class access `ImageSource.CancellationTokenSource` given that it's `private`?

**Comment by Ghostbird** [comment]
Similarly how can it access `IDispatcher.DispatchIfRequiredAsync` which is an `internal` function of `Microsoft.Maui.Controls`?

**Comment by bijington** [comment, marked as answer]
Hi @Ghostbird,

Please see my answers to your questions below:

> I still don't really get why it's done that way. 

I agree it is quite tricky to read. The underlying concept is to debounce any changes coming in and these can come from each of the properties on this class (e.g. `Email`, `DefaultImage`) as well as the the sizing changing of the control hosting the image. By debouncing we try to reduce the number of requests to gravatar.com.

> My questions are:
> * Should I do the same in my implementation?

I guess if it is likely that the stream could be requested multiple times in a very short space of time debouncing would be a good option.

> * How can the class access ImageSource.CancellationTokenSource given that it's private?

It is actually `private protected` meaning that we can access it through inheritance. See https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/private-protected for more details

> Similarly how can it access `IDispatcher.DispatchIfRequiredAsync` which is an `internal` function of `Microsoft.Maui.Controls`?

We have a special relationship with the .NET MAUI team and they expose their internals to this toolkit so that we can do such things. This is in the hope that tried and tested concepts in the toolkit can eventually be migrated into .NET MAUI itself.

**Reply by Ghostbird** [reply]
> The underlying concept is to debounce any changes coming in…

That the purpose was to debounce, is what I understood. What I don't understand is, why it is implemented in this way, if I reproduce it in my code, it doesn't work as a debounce at all. What I observed instead was: 

1. The class is instantiated
2. The `Source` is initiated
3. The `Uri` is initiated,this fires the `OnUriChanged()`
4. `OnUriChanged()`, immediately cancels the ongoing image load
5. `Source` function internally throws a `TaskCancelledException`
6. No image is ever loaded.

Of course my use case was slightly different, and I apparently cannot interface with MAUI the way this repo can, so I had to change some code. (e.g. my load isn't triggered by a `Uri` but by a string). I thought I had not, but I probably altered something that caused this behavioural change.

I have in my version added an early return that makes sure that the cancellation is skipped on changes where the `previousDispatch` is `null`, so that the initial load isn't immediately cancelled.

> We have a special relationship with the .NET MAUI team and they expose their internals to this toolkit so that we can do such things. This is in the hope that tried and tested concepts in the toolkit can eventually be migrated into .NET MAUI itself.

I guess my main confusion arose because the MAUI open source code apparently does not show the whole picture. Clearly a `private protected` member in MAUI cannot be accessed by another assembly. It is now clear to me that the community toolkit can do things with MAUI that cannot be done with the open source MAUI code on Github. I think that's quite annoying and confusing. Although I can understand the reason to an extent. I just wish it was more apparent before I started looking at community toolkit. When trying to implement my own classes, I often look at the community toolkit for examples, however it's clear now that's probably not a good idea, as the code here is mixed with _internal_ code that cannot work without direct integration into the MAUI assembly. The reason is, that it is at least partially intended for future integration.

**Reply by bijington** [reply]
What are you hoping to achieve with this implementation? Is it something that could benefit the toolkit and therefore be added to it and make use of the internals of MAUI?

**Reply by Ghostbird** [reply]
I need to load images from our server, which requires the underlying `HttpClient`  to be authenticated. Then we have some cache handling, and some special cases. We previously had this sprinkled across three classes, but ultimately it is a `StreamImageSource`. I've got it working now, but the images don't show up in MAUI yet.

I've verified that the `Stream` object is present, has the correct data, gets cached appropriately, gets reloaded from the cache appropriately and is correctly `Seek`ed to position 0 before the method returns. However, it's a different issue, as I've had similar issues when I did something subtly wrong. So it probably is not related to this specific class, but something I'm doing wrong in general when handling images in MAUI.

**Comment by CliffAgius** [comment]
Closed as answered
